### PR TITLE
feat(kuma-cp) domain name support in dataplane.networking.address

### PR DESCRIPTION
**Pull request by lobkovilya**
### Summary

On Universal deployment, a user has to manually set `networking.address` for Dataplane resource. It might be inconvenient since the address could be changed eventually. This PR gives an ability to set a domain name as a `networking.address`. The name will be resolved every time before xds-resource generation. By default names are cached with TTL equals to 10s.

### Full changelog

* domain name in `dataplane.networking.address`
* an ability to set dataplane-token via environment variable or command line argument.

### Issues resolved

Fix https://github.com/kumahq/kuma/issues/755 https://github.com/kumahq/kuma/issues/971

### Documentation

- [ ] Link to the website [documentation PR](https://github.com/kumahq/kuma-website/pull/XXX)

### CODE REVIEW

File: pkg/config/app/kuma-dp/config.go [hunk 1]
```diff
@@ -80,6 +80,8 @@ type DataplaneRuntime struct {
 	ConfigDir string `yaml:"configDir,omitempty" envconfig:"kuma_dataplane_runtime_config_dir"`
 	// Path to a file with dataplane token (use 'kumactl generate dataplane-token' to get one)
 	TokenPath string `yaml:"dataplaneTokenPath,omitempty" envconfig:"kuma_dataplane_runtime_token_path"`
+	// Token is dataplane token's value provided directly, will be store into the file in /tmp directory
```
Comment (nickolaev) [inline pkg/config/app/kuma-dp/config.go:83]: Can you rephrase it like this: `will be stored to a temporary file before applying`

File: pkg/core/dns/cache.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
+package dns
```
Comment (nickolaev) [inline pkg/core/dns/cache.go:1]: nit: Should the package be `dns_cache`?

File: pkg/core/resources/apis/mesh/dataplane_validator_test.go [hunk 3]
```diff
@@ -195,27 +195,6 @@ var _ = Describe("Dataplane", func() {
                 - field: networking
                   message: inbound cannot be defined both with gateway`,
 		}),
-		Entry("networking: invalid address", testCase{
```
Comment (nickolaev) [inline pkg/core/resources/apis/mesh/dataplane_validator_test.go:198]: Hmm, are we saying all stings are valid addresses now?

File: pkg/xds/topology/dataplanes.go [hunk 4]
```diff
@@ -3,21 +3,39 @@ package topology
...
+			return nil, err
+		}
 		if d.GetMeta().GetMesh() == mesh || d.Spec.IsIngress() {
 			_ = rv.AddItem(d)
 		}
 	}
 	return rv, nil
 }
+
+func ResolveAddress(dataplane *core_mesh.DataplaneResource) error {
```
Comment (nickolaev) [inline pkg/xds/topology/dataplanes.go:31]: Isn't this a strange place to put this function? Anything more generic?

File: pkg/core/resources/apis/mesh/dataplane_validator_test.go (hunk 3 above)
Comment (lobkovilya) [reply pkg/core/resources/apis/mesh/dataplane_validator_test.go:198]: I found regex for domain names, I think I can check against it:
`^([a-zA-Z0-9_]{1}[a-zA-Z0-9_-]{0,62}){1}(\.[a-zA-Z0-9_]{1}[a-zA-Z0-9_-]{0,62})*[\._]?$`

File: pkg/xds/topology/dataplanes.go (hunk 4 above)
Comment (lobkovilya) [reply pkg/xds/topology/dataplanes.go:31]: Probably yes, but it depends on `dataplane` and used only inside xds package. Do you have any suggestions about another place?

File: pkg/core/resources/apis/mesh/dataplane_validator_test.go (hunk 3 above)
Comment (nickolaev) [reply pkg/core/resources/apis/mesh/dataplane_validator_test.go:198]: https://www.golangprograms.com/regular-expression-to-extract-dns-host-name-or-ip-address-from-string.html

Comment (lobkovilya) [reply pkg/core/resources/apis/mesh/dataplane_validator_test.go:198]: There is only IP address in the provided example

File: pkg/core/dns/cache.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+func MakeCaching(f LookupIPFunc, ttl time.Duration) LookupIPFunc {
+	cache := map[string]*cacheRecord{}
+	var rwmux sync.RWMutex
+	return func(host string) ([]net.IP, error) {
+		rwmux.RLock()
+		r, ok := cache[host]
+		rwmux.RUnlock()
+
+		if ok && r.creationTime.Add(ttl).After(time.Now()) {
```
Comment (jakubdyszkiewicz) [inline pkg/core/dns/cache.go:24]: `core.Now()`

File: pkg/core/dns/cache.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+			return r.ips, nil
+		}
+
+		ips, err := f(host)
+		if err != nil {
+			return nil, err
+		}
+
+		rwmux.Lock()
+		cache[host] = &cacheRecord{ips: ips, creationTime: time.Now()}
```
Comment (jakubdyszkiewicz) [inline pkg/core/dns/cache.go:34]: `core.Now()`

File: pkg/core/dns/cache_test.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+		table["example.com"] = []net.IP{net.ParseIP("192.168.0.1")}
+
+		ip, _ := cachingLookupFunc("example.com")
+		Expect(ip[0]).To(Equal(net.ParseIP("192.168.0.1")))
+
+		ip, _ = cachingLookupFunc("example.com")
+		Expect(ip[0]).To(Equal(net.ParseIP("192.168.0.1")))
+
+		table["example.com"] = []net.IP{net.ParseIP("10.20.0.1")}
+		<-time.After(2 * time.Second)
```
Comment (jakubdyszkiewicz) [inline pkg/core/dns/cache_test.go:44]: instead, just change `core.Now` so we avoid adding extra 2 seconds of waiting to our tests

File: pkg/core/resources/apis/mesh/dataplane_validator_test.go (hunk 3 above)
Comment (jakubdyszkiewicz) [reply pkg/core/resources/apis/mesh/dataplane_validator_test.go:198]: I bet someone will try to put there `http://` or port. Please validate that it is a hostname or ip

File: pkg/xds/topology/dataplanes.go [hunk 8]
```diff
@@ -3,21 +3,39 @@ package topology
...
+	ips, err := core.LookupIP(dataplane.Spec.Networking.Address)
```
Comment (jakubdyszkiewicz) [inline pkg/xds/topology/dataplanes.go:32]: so what happens when it is IP (which is like 99% of environments)? Do we execute syscall trying to resolve IP? Can we have an if here that if the address is an IP then we skip this step?

File: pkg/core/alias.go [hunk 9]
```diff
@@ -15,6 +18,7 @@ var (
 	NewLogger = kuma_log.NewLogger
 	SetLogger = kube_log.SetLogger
 	Now       = time.Now
+	LookupIP  = dns.MakeCaching(net.LookupIP, 10*time.Second)
```
Comment (jakubdyszkiewicz) [inline pkg/core/alias.go:21]: I'd like to avoid having it in `core`. This list of aliases should be used only for essential-critical-core stuff.
My suggestion, place it in the file with cache

Also by doing this I think you just dragged DNS server dependency on anyone who is importing `core` even just for logging (kuma-dp)

Comment (jakubdyszkiewicz) [inline pkg/core/alias.go:21]: please parametrize this 10 seconds. I have a feeling this will be useful value to change while debugging stuff

File: pkg/xds/topology/dataplanes.go (hunk 4 above)
Comment (jakubdyszkiewicz) [inline pkg/xds/topology/dataplanes.go:31]: Overall, I need some explanation of why are we resolving an IP address. Please write a doc stating that EDS requires IP, not hostname but for some environments, it's much more convenient to have hostname in address.

File: pkg/xds/topology/dataplanes.go [hunk 10]
```diff
@@ -3,21 +3,39 @@ package topology
...
+	if err != nil {
+		return err
+	}
+	if len(ips) == 0 {
+		return errors.Errorf("can't resolve address %v", dataplane.Spec.Networking.Address)
+	}
+	dataplane.Spec.Networking.Address = ips[0].String()
```
Comment (jakubdyszkiewicz) [inline pkg/xds/topology/dataplanes.go:39]: please double-check out all usages of Address. I see `GetIP()` method on Dataplane and we use in Kuma Prometheus SD. What will happen there if we suddenly return hostname, and not IP?

File: pkg/core/dns/cache.go [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+type LookupIPFunc func(string) ([]net.IP, error)
+
+type cacheRecord struct {
+	ips          []net.IP
+	creationTime time.Time
+}
+
+func MakeCaching(f LookupIPFunc, ttl time.Duration) LookupIPFunc {
```
Comment (jakubdyszkiewicz) [inline pkg/core/dns/cache.go:16]: I think we can figure out better name. Maybe `CachedLookupIP`? `CachedLookupIPFunc`?

File: pkg/core/dns/cache.go (hunk 2 above)
Comment (jakubdyszkiewicz) [reply pkg/core/dns/cache.go:1]: I think Nikolay is kind of right here. I'd like to avoid confusion that this is part of DNS server. How about `dns/lookup` package with `lookup.go` (or `interface.go`) and `cache.go` files?

File: pkg/xds/topology/dataplanes.go (hunk 4 above)
Comment (jakubdyszkiewicz) [reply pkg/xds/topology/dataplanes.go:31]: I like this one. I don't like having half of the project in `utils` directory, because it _may_ be used somewhere

File: app/kuma-dp/cmd/run.go [hunk 12]
```diff
@@ -135,5 +147,13 @@ func newRunCmd() *cobra.Command {
 	cmd.PersistentFlags().StringVar(&cfg.DataplaneRuntime.BinaryPath, "binary-path", cfg.DataplaneRuntime.BinaryPath, "Binary path of Envoy executable")
 	cmd.PersistentFlags().StringVar(&cfg.DataplaneRuntime.ConfigDir, "config-dir", cfg.DataplaneRuntime.ConfigDir, "Directory in which Envoy config will be generated")
 	cmd.PersistentFlags().StringVar(&cfg.DataplaneRuntime.TokenPath, "dataplane-token-file", cfg.DataplaneRuntime.TokenPath, "Path to a file with dataplane token (use 'kumactl generate dataplane-token' to get one)")
+	cmd.PersistentFlags().StringVar(&cfg.DataplaneRuntime.Token, "dataplane-token", cfg.DataplaneRuntime.Token, "Dataplane Token")
 	return cmd
 }
+
+func writeFile(filename string, data []byte, perm os.FileMode) error {
+	if err := os.MkdirAll(filepath.Dir(filename), 0755); err != nil {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply app/kuma-dp/cmd/run.go:155]: TBH I just copied this little function from there https://github.com/kumahq/kuma/blob/master/app/kuma-dp/pkg/dataplane/envoy/bootstrap_file.go#L33, and I believe it makes sense to have 755 for directories to be able to `cd` into them

File: pkg/xds/topology/dataplanes.go (hunk 8 above)
Comment (lobkovilya) [reply pkg/xds/topology/dataplanes.go:32]: `net.LookupIP` tries to parse IP address inside - https://golang.org/src/net/lookup.go?s=6141:6181#L242 and return

File: pkg/xds/topology/dataplanes.go (hunk 4 above)
Comment (nickolaev) [reply pkg/xds/topology/dataplanes.go:31]: What if it is a `core_mesh.DataplaneResource` method?

File: app/kuma-dp/cmd/run.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply app/kuma-dp/cmd/run.go:155]: this is a dataplane token, it should not be read by any other process than Kuma.

File: pkg/xds/topology/dataplanes.go (hunk 4 above)
Comment (jakubdyszkiewicz) [reply pkg/xds/topology/dataplanes.go:31]: and then you are introducing DNS dependency into `core_mesh` package. I think `core` packages should not depend on other packages next to core (aside of a couple of exceptions like `api` and `utils`), but as long as Ilya will move it to a separate package so we don't drag the whole DNS server into `core_mesh` I can live with it for now ;)

File: pkg/core/dns/cache.go (hunk 2 above)
Comment (lobkovilya) [reply pkg/core/dns/cache.go]: Fixed, create package `lookup` in `dns` dir and 2 files: `cache.go` and `interface.go`

File: pkg/core/resources/apis/mesh/dataplane_validator_test.go (hunk 3 above)
Comment (lobkovilya) [reply pkg/core/resources/apis/mesh/dataplane_validator_test.go:198]: Fixed, validate using regex

File: pkg/xds/topology/dataplanes.go (hunk 10 above)
Comment (lobkovilya) [reply pkg/xds/topology/dataplanes.go:40]: I resolve the address as soon as dataplanes are fetched. Only one place has an unresolved host - MADS, but according to the docs https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config, it can accept `host:port` format

File: pkg/core/alias.go (hunk 9 above)
Comment (lobkovilya) [reply pkg/core/alias.go]: do you mean to put this value into Config? Then I probably have to move LookupIP initialization into RuntimeBuilder or something like that, is it okay?

File: pkg/xds/topology/dataplanes.go (hunk 4 above)
Comment (lobkovilya) [reply pkg/xds/topology/dataplanes.go:32]: Do you mean doc as a comment right in the code? Do you think makes sense to mention ECS in this place in the code?

Comment (lobkovilya) [reply pkg/xds/topology/dataplanes.go]: Added info about reasons

File: pkg/core/alias.go (hunk 9 above)
Comment (lobkovilya) [reply pkg/core/alias.go]: Added new parameter - `Config.GeneralConfig.DNSCacheTTL`, not sure about place for this parameter in the config

File: pkg/xds/topology/dataplanes.go [hunk 13]
```diff
@@ -3,23 +3,22 @@ package topology
...
 	dataplanes := &core_mesh.DataplaneResourceList{}
-	if err := rm.List(ctx, dataplanes); err != nil {
+	if err := rt.ReadOnlyResourceManager().List(ctx, dataplanes); err != nil {
 		return nil, err
 	}
 	rv := &core_mesh.DataplaneResourceList{}
 	for _, d := range dataplanes.Items {
-		if err := ResolveAddress(d); err != nil {
+		if err := ResolveAddress(rt, d); err != nil {
 			return nil, err
```
Comment (jakubdyszkiewicz) [inline pkg/xds/topology/dataplanes.go:22]: we should log and ignore such dataplane. Otherwise one dataplane with address that cannot be resolved will break the entire mesh

File: pkg/xds/topology/dataplanes.go [hunk 14]
```diff
@@ -3,23 +3,22 @@ package topology
...
 
 	"github.com/pkg/errors"
 
 	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
-	"github.com/kumahq/kuma/pkg/core/resources/manager"
 )
 
 // GetDataplanes returns list of Dataplane in provided Mesh and Ingresses (which are cluster-scoped, not mesh-scoped)
-func GetDataplanes(ctx context.Context, rm manager.ReadOnlyResourceManager, mesh string) (*core_mesh.DataplaneResourceList, error) {
+func GetDataplanes(rt runtime.RuntimeContext, ctx context.Context, mesh string) (*core_mesh.DataplaneResourceList, error) {
```
Comment (jakubdyszkiewicz) [inline pkg/xds/topology/dataplanes.go:14]: nit: I'd rather have GetDataplanes to depend on `ReadOnlyResourceManager` and `LookupIP`. Dependencies of the functions are then clear. Otherwise you need to figure out what exactly is picked up from RuntimeContext. Also it should be easier to test then

File: pkg/xds/server/components.go [hunk 15]
```diff
@@ -131,7 +131,7 @@ func DefaultDataplaneSyncTracker(rt core_runtime.Runtime, reconciler, ingressRec
 						return err
 					}
 					for _, dp := range allMeshDataplanes.Items {
-						if err := xds_topology.ResolveAddress(dp); err != nil {
+						if err := xds_topology.ResolveAddress(rt, dp); err != nil {
```
Comment (jakubdyszkiewicz) [inline pkg/xds/server/components.go:134]: same remark here about and ignoring DP

File: pkg/xds/topology/dataplanes.go [hunk 16]
```diff
@@ -29,8 +28,11 @@ func GetDataplanes(ctx context.Context, rm manager.ReadOnlyResourceManager, mesh
 	return rv, nil
 }
 
-func ResolveAddress(dataplane *core_mesh.DataplaneResource) error {
-	ips, err := dns.LookupIP(dataplane.Spec.Networking.Address)
+// ResolveAddress resolves 'dataplane.networking.address' if it has DNS name in it. This is a crucial feature for
+// some environments specifically AWS ECS. Dataplane resource has to be created before running Kuma DP, but IP address
+// will be assigned only after container's start. Being able to set dp's address as a DNS name solves this problem
```
Comment (jakubdyszkiewicz) [inline pkg/xds/topology/dataplanes.go:33]: yeah, but then again - why are we resolving? :)

Add info that "Hostname cannot be used in EDS"

File: pkg/config/app/kuma-cp/config.go [hunk 17]
```diff
@@ -249,6 +251,8 @@ type GeneralConfig struct {
 	// Hostname that other components should use in order to connect to the Control Plane.
 	// Control Plane will use this value in configuration generated for dataplanes, in responses to `kumactl`, etc.
 	AdvertisedHostname string `yaml:"advertisedHostname" envconfig:"kuma_general_advertised_hostname"`
+	// DNSCacheTTL represents duration for how long Kuma CP will cache result of resolving dataplane's domain name
+	DNSCacheTTL time.Duration `yaml:"dnsCacheTTL" envconfig:"kuma_general_dns_cache_ttl"`
```
Comment (jakubdyszkiewicz) [inline pkg/config/app/kuma-cp/config.go:255]: I'm hesitant whether this should be placed in `general`, `xds` for me seems more appropriate, but I don't have strong opinion about this, so I'll let you choose

**Comment by subnetmarco** [comment]
Do we need to update the docs as a result of this PR?

**Comment by lobkovilya** [comment]
Yes, right, I will update it
